File: java/CoinChange2.java
class Solution {
    public int change(int amount, int[] coins) {
        int[] res = new int[amount + 1];
        res[0] = 1;
        // dp[i, j] = dp[i - 1, j] + dp[i, j - coins[i]]
        for (int i = 0; i < coins.length; i++){
            for (int j = coins[i]; j <= amount; j++){
                res[j] = res[j] + res[j - coins[i]];
            }
        }
        
        return res[amount];
    }
}

File: python/CoinChange2.py
class Solution(object):
    def change(self, amount, coins):
        """
        :type amount: int
        :type coins: List[int]
        :rtype: int
        """
        
        res = [1] + [0] * amount
        for coin in coins:
            for i in range(coin, amount + 1):
                res[i] = res[i] + res[i - coin]
        return res[amount]
